New Feature Multiple Ground Temperature and Ground Reflectance

Nigusse opened the pull request:
Pull request overview
---------------------
 - New Feature
 - Allows multiple ground surface temperature specification for use with LWR exchange calculation with exterior surface and multiple ground surface reflectance for ground reflected solar radiation calculation for use with exterior surfaces heat balance.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

Nigusse committed:
NFP-MultipleGroundTemperatureReflectance

Nigusse committed:
NFP - Initial Upload Cleanup

Nigusse committed:
NFP - Initial Upload Cleanup

Nigusse committed:
NFP Cleanup

Nigusse committed:
Merge branch 'develop' into MultipleGroundTemperatureReflectance

hongtz68 commented:
@Nigusse This is a recap from today's technicalities call:

1. This feature requests to enable users to specify multiple ground surfaces to account for their longwave radiant heat exchange with exterior surfaces of buildings.
2. Each ground surface should have its own view factor, solar reflectance, and temperature (as a schedule value).
3. Surrounding surfaces participating in the longwave radiant heat exchange do not reflect solar. This is the current EnergyPlus function, we don't want to change it as the solar reflectance calculation would need more information including reflectance, orientation, and shading of each surrounding surface. To account for solar reflectance from surrounding surfaces, these surfaces can be defined as shading surfaces, which is an existing feature of EnergyPlus.
4. In the future, we can consolidate surrounding surfaces and exterior/site shading surfaces.

hongtz68 commented on design/FY2022/NFP-MultipleGroundTemperatureReflectance.md:
```diff
@@ -0,0 +1,183 @@
+       \type object-list
+       \object-list ScheduleNames
+       \note Schedule values are real numbers, -100.0 to 100.0, units C	   
+`   A7, \field Surrounding Surface 1 Reflectance Schedule Name
```
we don't need this field.

Nigusse replied:
OK.

hongtz68 commented on design/FY2022/NFP-MultipleGroundTemperatureReflectance.md:
```diff
@@ -0,0 +1,183 @@
+`       \required-field
+`       \type object-list
+`       \object-list ScheduleNames
+`       \note Schedule values are fractions, 0.0 to 1.0, dimensionless
```
need to allow multiple ground surfaces.

Nigusse replied:
@hongtz68 So, if I understand correctly, what you mean is that one Ground Reflectance Schedule can be specified for each exterior surface. Also if needed the same reflectance schedule can be used in multiple exterior surface, for example exterior surface with the same orientation. Is this correct interpretation of your intent?

Nigusse commented:
> @Nigusse This is a recap from today's technicalities call:
> 

@hongtz68 Thank you for the summary.

hongtz68 commented:
That is not what I meant. Surrounding surfaces include ground surfaces (currently only one is allowed and this is what we try to address) and exterior surfaces of nearby buildings. This feature request IS to allow multiple ground surfaces, nothing needs to be changed for other types of surrounding surfaces (i.e., exterior surfaces of nearby buildings). 

Now with multiple ground surfaces, we need to allow each ground surface has its own view factor, reflectance and temperature. As said already, we don't consider solar reflectance from surrounding surfaces of nearby buildings in this feature.

Nigusse commented:
@hongtz68 Thank you for the clarification. OK, here is an approach to consider. I propose a new object ***SurfaceProperty:GroundSurfaces*** that holds multiple ground surfaces properties and specify the new object name in **SurfaceProperty:LocalEnvironment** object by adding a **NEW FIELD 6** called ***Ground Surfaces Object Name***. 

See below a modified existing objects.

```
SurfaceProperty:LocalEnvironment,
   A1, \field Name
       \required-field
       \type alpha
       \reference SurfaceLocalEnvironmentNames
   A2, \field Exterior Surface Name
       \type object-list
       \object-list SurfaceNames
       \note Enter the name of an exterior surface object
       ...
   A5, \field Outdoor Air Node Name
       \type object-list
       \object-list OutdoorAirNodeNames
       \note Enter the name of an OutdoorAir:Node object
   A6; \field Ground Surfaces Object Name
       \type object-list
       \object-list GroundSurfacesNames
       \note Enter the name of a SurfaceProperty:GroundSurfaces object  
```

Proposed New Object:

```
SurfaceProperty:GroundSurfaces,
       \min-fields 5
       \memo This object defines a list of ground surfaces for use with an exterior surface.
       \extensible:4 -- duplicate last set of ground surface properties (the last four fields).
   A1, \field Name
       \required-field
       \type alpha
       \reference SurroundingSurfacesNames
   A2, \field Ground Surface 1 Name
       \begin-extensible
       \required-field
       \type alpha
   N1, \field Ground Surface 1 View Factor
       \required-field
       \minimum 0.0
       \maximum 1.0
       \default 0.5
   A3, \field Ground Surface 1 Temperature Schedule Name
       \required-field
       \type object-list
       \object-list ScheduleNames
       \note Schedule values are real numbers, -100.0 to 100.0, units C
   A4, \field Ground Surface 1 Reflectance Schedule Name
       \required-field
       \type object-list
       \object-list ScheduleNames
       \note Schedule values are fraction, 0.0 to 1.0, units dimensionless
       
      extensible fields
       '''
```

The following two fields in "SurfaceProperty:SurroundingSurfaces," object will be deprecated if we decide to use this approach.

```
"SurfaceProperty:SurroundingSurfaces,"
  ...
  N2, \field Ground View Factor
       \minimum 0.0
       \maximum 1.0
       \default 0.5
       \note optional
   A3, \field Ground Temperature Schedule Name
       \type object-list
       \object-list ScheduleNames
       \note Schedule values are real numbers, -100.0 to 100.0, units C
       \note optional
```

hongtz68 commented:
This works and provides flexibility.

Nigusse commented:
@hongtz68 Thanks. I will go ahead update the NFP accordingly.

Nigusse committed:
Updated NFP (PR #9409)

Nigusse committed:
Merge branch 'develop' into MultipleGroundTemperatureReflectance

Nigusse committed:
Modified Idd

Nigusse committed:
getInput for new object and new field

Nigusse committed:
Calc Average GndTemp Function

Nigusse committed:
apply clang format

Nigusse committed:
Set Ground Temp for outside surface heat balance

Nigusse committed:
Multiple Ground Surface Reflectance

Nigusse committed:
Updated Ground Surface Temp Use

Nigusse committed:
Merge branch 'develop' into MultipleGroundTemperatureReflectance

Nigusse commented:
Added Design Document to the NFP (PR #9409)

Nigusse committed:
Merge branch 'MultipleGroundTemperatureReflectance' into MultGndTempTest

Nigusse committed:
Reset GroundViewFactorIR in solar shading module

Nigusse committed:
Modified IDD and Reset GroundViewFactorIR

Nigusse committed:
Merge branch 'develop' into MultGndTempTest

Nigusse committed:
Removed two fields from SurroundingSurfaces obj

Nigusse committed:
modified GetSurfaceSrdSurfsData() function

Nigusse committed:
Removed ground VF and ground temp vars

Nigusse committed:
Transitioned unit tests and fixed failed unit tests

Nigusse committed:
Transitioned an example idf

Nigusse committed:
updated where ground reflectance is applied

Nigusse committed:
Merge branch 'develop' into MultipleGroundTemperatureReflectance

Nigusse committed:
Merge branch 'MultipleGroundTemperatureReflectance' into MultGndTempTest

Nigusse committed:
cleanup

Nigusse committed:
Added Unit Test

Nigusse commented:
The NFP has been updated based on the feedback I received in the technical call. Also I have added design document for implementation. I am waiting for any feedback before starting implementation. @Myoldmopar @mjwitte @hongtz68

Nigusse committed:
More Unit Test

Nigusse committed:
Again More Unit Test

Nigusse committed:
More Unit Tests

Nigusse committed:
Merge branch 'develop' into MultGndTempTest

Nigusse committed:
Modified View Factors Update

hongtz68 commented:
The revised NFP and the added design document look good.

Nigusse committed:
Merge branch 'develop' into MultipleGroundTemperatureReflectance

Nigusse committed:
Merge branch 'MultipleGroundTemperatureReflectance' into MultGndTempTest

Nigusse committed:
More Cleanup

Nigusse committed:
Modified GetInputs and Cleanup

Nigusse committed:
Various code cleanup

Nigusse committed:
Updated NFP IO and Eng Ref

Nigusse committed:
Modified Surrounding Surfaces Object IO Ref Docs

Nigusse committed:
Removed Ground Surface Fields from Surrounding Surfaces Object IO Ref Docs

Nigusse committed:
Added Ground Surfaces Object IO Ref Docs

Nigusse committed:
Modified Local Environment Object IO Ref Docs

Nigusse committed:
IO Ref Docs Cleanup

Nigusse committed:
Eng Ref Docu for LWR of Multilpe Gnd Surfaces

Nigusse committed:
Eng Ref Docs for Refl of Multiple Gnd Surfaces

Nigusse committed:
Merge branch 'MultipleGroundTemperatureReflectance' into MultGndTempTest

Nigusse committed:
Modified GroundSurfaces GetInputs

Nigusse committed:
Ground Surface 1 View Factor Field Not Required

Nigusse committed:
Modified Unit Test

Nigusse committed:
Removed default for Ground Surface 1 View Factor field

Nigusse committed:
Added New Example File

Nigusse committed:
Reduced Minimum Field for GroundSurfaces

Nigusse committed:
Modified Ground Surface View Factor Reset

Nigusse committed:
Modified the new example file

Nigusse committed:
Corrected alpha field numbering

Nigusse committed:
Modified View Factor Reset Calc

Nigusse committed:
Added more unit test

Nigusse committed:
New Unit Tests Cleanup

Nigusse committed:
Merge branch 'develop' into MultipleGroundTemperatureReflectance

Nigusse committed:
Merge branch 'MultipleGroundTemperatureReflectance' into MultGndTempTest

Nigusse committed:
Updated Gnd Reflectance Use

Nigusse committed:
Gnd Surf Reflected Solar Rad Calc by Surface

Nigusse committed:
Removed Default Constructor

Nigusse committed:
Merge branch 'develop' into MultipleGroundTemperatureReflectance

Nigusse committed:
Merge branch 'MultipleGroundTemperatureReflectance' into MultGndTempTest

Nigusse committed:
Initial Code Changes and Transition Rule

Nigusse committed:
Fixed Clang Format Error

Nigusse committed:
Removed tabs from new example file

Nigusse committed:
Fixed unused variable warning message

Nigusse committed:
Addressed Warnings and Merged in Develop

Nigusse committed:
Removed tab from example file

Nigusse committed:
Modified Ground Surface Prop Calc

Nigusse commented on doc/engineering-reference/src/solar-radiation-reflected-from-exterior/diffuse-reflection-of-beam-solar-and-sky.tex:
```diff
@@ -130,6 +130,32 @@ \subsection{Sky Solar Radiation Diffusely Reflected from the Ground}\label{sky-s
+
+\({\rho_{gnd,avg}}\) = view factor weighed ground solar reflectance of multiple ground surfaces viewed by an exterior surface
+
+
```
Updates eng reference documentation.

Nigusse commented on doc/engineering-reference/src/solar-radiation-reflected-from-exterior/diffuse-reflection-of-beam-solar-and-sky.tex:
```diff
@@ -192,6 +218,33 @@ \subsection{Beam Solar Radiation Diffusely Reflected from the Ground}\label{beam
+
+\({\rho_{gnd,avg}}\) = view factor weighed ground solar reflectance of multiple ground surfaces viewed by an exterior surface
+
+
```
Updates eng reference documentation.

Nigusse commented on doc/engineering-reference/src/surface-heat-balance-manager-processes/outside-surface-heat-balance.tex:
```diff
@@ -169,6 +169,53 @@ \subsection{External Longwave Radiation}\label{external-longwave-radiation}
+
+h\(_{r,gnd,avg}\) = linearized average radiative heat transfer coefficient between an exterior surface and multiple ground surfaces
+
+
```
Updates eng reference documentation.

Nigusse commented on doc/input-output-reference/src/overview/group-advanced-surface-concepts.tex:
```diff
@@ -2519,11 +2519,11 @@ \subsubsection{Inputs}\label{inputs-16}
 
-The object also defines the sky and ground temperature and view factors to the external surface. The sum of all defined view factors should not exceed 1.0. If only sky view is defined in this object, the ground view factor to this surface will be 1.0 subtracted with the ground view factor and all other defined surface view factors. If only ground view is defined in this object, the sky view facto ...
+The object also defines the sky temperature and view factors to the external surface. The sum of all defined view factors including the ground surfaces view factors defined in \textit{SurfaceProperty:GroundSurfaces} object should not exceed 1.0. If sky view factor is not defined in this object, the sky view factor to this surface will be 1.0 subtracted the ground surface view factors defined in \ ...
 
```
Updated IO Reference docs.

Nigusse commented on doc/input-output-reference/src/overview/group-advanced-surface-concepts.tex:
```diff
@@ -2537,14 +2537,6 @@ \subsubsection{Field: Sky Temperature Schedule Name}\label{field-sky-temperature
-\subsubsection{Field: Ground Temperature Schedule Name}\label{field-ground-temperature-schedule-name}
-
-This field is used to supply a schedule name of the ground temperature. The ground temperature used in long wave radiation exchange calculation of the surface would be overwritten with the value of the scheduled temperature if defined here. If the field is left blank, the global ground temperature would be used.
-
```
Deleted fields. Replaced with a new object `SurfaceProperty:GroundSurfaces` that allows to specify multiple ground surfaces properties.

Nigusse commented on doc/input-output-reference/src/overview/group-advanced-surface-concepts.tex:
```diff
@@ -2580,9 +2570,95 @@ \subsubsection{Field: Surrounding Surface 1 Temperature Schedule Name}\label{fie
+
+Bill Marion. 2018. Ground Albedo Measurements and Modeling. Bifacial PV Workshop Lakewood, Colorado. September 11, 2018.
+NREL/PR-5K00-72589. https://www.nrel.gov/docs/fy20osti/72589.pdf
+
```
IO Reference documentation for the new object `SurfaceProperty:GroundSurfaces`.

Nigusse commented on doc/input-output-reference/src/overview/group-advanced-surface-concepts.tex:
```diff
@@ -2598,12 +2674,16 @@ \subsubsection{Field: External Shading Fraction Schedule Name}\label{field-exter
+\subsubsection{Field: Ground Surfaces Object Name}\label{field-ground-surfaces-object-name}
+
+This field is used to import ground surfaces properties (ground surface temperature, solar reflectance and view factors to the external surface) from external calculation. This should be a name of a \textbf{\hyperref[surfacePropertyGroundSurfaces]{SurfaceProperty:GroundSurfaces}} object.
+
```
New field added to `SurfaceProperty:LocalEnvironment` object. The new field is used to enter the name of the `SurfaceProperty:GroundSurfaces` object.

Nigusse commented on idd/Energy+.idd.in:
```diff
@@ -18632,10 +18632,15 @@ SurfaceProperty:LocalEnvironment,
+       \type object-list
+       \object-list GroundSurfacesNames
+       \note Enter the name of a SurfaceProperty:GroundSurfaces object
+	   \note This field is used with SurfaceProperty:GroundSurfaces object
```
This text needs correction. ~\note This field is used with SurfaceProperty:GroundSurfaces object~.

Nigusse commented on idd/Energy+.idd.in:
```diff
@@ -18655,7 +18660,7 @@ ZoneProperty:LocalEnvironment,
 
 SurfaceProperty:SurroundingSurfaces,
-       \min-fields 8
+       \min-fields 6
```
Removed two fields. Thus reduced the min # of fields.

Nigusse commented on idd/Energy+.idd.in:
```diff
@@ -18672,120 +18677,217 @@ SurfaceProperty:SurroundingSurfaces,
        \object-list ScheduleNames
        \note Schedule values are real numbers, -100.0 to 100.0, units C
        \note optional
-   N2, \field Ground View Factor
```
Removed two fields from `SurfaceProperty:SurroundingSurfaces` object.

mjwitte replied:
@Nigusse I did not realize until now that these fields are not present in the new SurfaceProperty:GroundSurfaces object. This eliminates the current method of simply assigning a ground view factor without describing ground surfaces. Perhaps these fields should remain here, but be ignored if these is a SurfaceProperty:GroundSurfaces object referenced here.

Nigusse replied:
@mjwitte I will do if it is worth it but I have to make several changes to retain the current method.

mjwitte replied:
@Nigusse Think through the options, as-is, keep the ground fields here, or keep them in SurfaceProperty:GroundSurfaces. With the current design, how do you transition and get the same results for a file that had, say, 0.3 for Ground View Factor in the original file?

Nigusse replied:
@mjwitte I rolled back the two deleted fields per your original suggestion. Also rolled back the transitioned example file `SurfacePropTest_SurfLWR.idf`. The transitioned and the current idf are expected to give identical results. Local test produced identical results and also previous regression run (prior the roll back) did not produce any diffs.

Nigusse replied:
> With the current design, how do you transition and get the same results for a file that had, say, 0.3 for Ground View Factor in the original file?

@mjwitte Was there any reason why you thought the original results could not be duplicated with the new method?

mjwitte replied:
> > With the current design, how do you transition and get the same results for a file that had, say, 0.3 for Ground View Factor in the original file?
> 
> @mjwitte Was there any reason why you thought the original results could not be duplicated with the new method?

@Nigusse That question applied to the original IDD design where "Ground View Factor" was not an available field.

Nigusse replied:
OK. The transition created the new object with the view factor value from the deleted fields.

mjwitte replied:
Yes, but a view factor to what? There had to be a surface name, correct? What surface represent the general ground?

Nigusse replied:
Below is the current example file surrounding surfaces object:
```
  SurfaceProperty:SurroundingSurfaces,
    SrdSurfs:Zn005:Wall002:Win001,  !- Name
    0.5,                     !- Sky View Factor
    ,                        !- Sky Temperature Schedule Name
    0.3,                     !- Ground View Factor
    ,                        !- Ground Temperature Schedule Name
    SurroundingSurface1,     !- Surrounding Surface 1 Name
    0.2,                     !- Surrounding Surface 1 View Factor
    Surrounding Temp Sch 2;  !- Surrounding Surface 1 Temperature Schedule Name
```

Below is the complete transitioned objects. The last new field `Ground Surfaces Object Name` added to the `SurfaceProperty:LocalEnvironment` object links the new ground surface object to the exterior surface, just like the surrounding surfaces object is referenced. The transition rule was able to do this:
 
```
  SurfaceProperty:SurroundingSurfaces,
    SrdSurfs:Zn005:Wall002:Win001,  !- Name
    0.5,                     !- Sky View Factor
    ,                        !- Sky Temperature Schedule Name
    SurroundingSurface1,     !- Surrounding Surface 1 Name
    0.2,                     !- Surrounding Surface 1 View Factor
    Surrounding Temp Sch 2;  !- Surrounding Surface 1 Temperature Schedule Name

 SurfaceProperty:GroundSurfaces,
    GndSurfs:Zn005:Wall002:Win001,  !- Name
    Gnd Surfs Wall002:Win001 Name,  !- Ground Surface 1 Name
    0.3;                     !- Ground Surface 1 View Factor

  SurfaceProperty:LocalEnvironment,
    LocEnv:Zn005:Wall002:Win001,  !- Name
    Zn005:Wall002:Win001,    !- Exterior Surface Name
    ,                        !- External Shading Fraction Schedule Name
    SrdSurfs:Zn005:Wall002:Win001,  !- Surrounding Surfaces Object Name
    ,                        !- Outdoor Air Node Name
   *GndSurfs:Zn005:Wall002:Win001;  !- Ground Surfaces Object Name*

```

mjwitte replied:
But what is this? `    Gnd Surfs Wall002:Win001 Name,  !- Ground Surface 1 Name`

Nigusse replied:
Note that the `SurfaceProperty:GroundSurfaces` main intent was to allow multiple ground surfaces. So each ground surface will have its own surface name. This was clearly defined in the IDD. See an example below:

```
  SurfaceProperty:GroundSurfaces,
    GndSurfs:Zn003:Wall001,  !- Name

    GndSurfs:Grass,          !- Ground Surface 1 Name
    0.2,                     !- Ground Surface 1 View Factor
    GndSurfs:GrassTemp,      !- Ground Surface 1 Temperature Schedule Name
    GndSurfs:GrassRefl,      !- Ground Surface 1 Reflectance Schedule Name

    GndSurfs:Parking,        !- Ground Surface 2 Name
    0.1,                     !- Ground Surface 2 View Factor
    GndSurfs:ParkingTemp,    !- Ground Surface 2 Temperature Schedule Name
    GndSurfs:ParkingRefl,    !- Ground Surface 2 Reflectance Schedule Name

    GndSurfs:Pond,           !- Ground Surface 3 Name
    0.2,                     !- Ground Surface 3 View Factor
    GndSurfs:PondTemp,       !- Ground Surface 3 Temperature Schedule Name
    GndSurfs:PondRefl;       !- Ground Surface 3 Reflectance Schedule Name
```

Nigusse replied:
The above object snippet was taken from the new example file `SurfacePropGroundSurfaces_LWR.idf`.

mjwitte replied:
> Note that the `SurfaceProperty:GroundSurfaces` main intent was to allow multiple ground surfaces. So each ground surface will have its own surface name.

Understood. The transitioned example had this object:
```
SurfaceProperty:GroundSurfaces,
    GndSurfs:Zn005:Wall002:Win001,  !- Name
    Gnd Surfs Wall002:Win001 Name,  !- Ground Surface 1 Name
    0.3;                     !- Ground Surface 1 View Factor
```
But the file has no surface named " Gnd Surfs Wall002:Win001 Name".

Nigusse replied:
Of course there is no "ground surface" named surface object. This is just a name that identifies one of the ground surfaces viewed by an exterior surface. When an exterior surface views multiple ground surfaces, it it is good to have an identifier name for each ground surface with different properties.

Nigusse commented on idd/Energy+.idd.in:
```diff
@@ -18672,120 +18677,217 @@ SurfaceProperty:SurroundingSurfaces,
        \maximum 1.0
-       \default 0.5
-       \note optional
-   A3, \field Ground Temperature Schedule Name
```
The fields `Surrounding Surface 1 View Factor` is initialized to -1.0 if left blank. No need to have a default value.

Nigusse commented on idd/Energy+.idd.in:
```diff
@@ -18672,120 +18677,217 @@ SurfaceProperty:SurroundingSurfaces,
+	   \units dimensionless
+       \minimum 0.0
+       \maximum 1.0
+   A3, \field Ground Surface 1 Temperature Schedule Name
```
The new idd object `SurfaceProperty:GroundSurfaces`.

Nigusse commented on src/EnergyPlus/ConvectionCoefficients.cc:
```diff
@@ -415,10 +415,10 @@ void InitExteriorConvectionCoeff(EnergyPlusData &state,
+    if (state.dataSurface->UseSurfPropertyGndSurfTemp(SurfNum)) {
+        int gndSurfsNum = state.dataSurface->GroundSurfsPropertyNum(SurfNum);
+        TGround = state.dataSurface->GroundSurfsProperty(gndSurfsNum).SurfsTempAvg + DataGlobalConstants::KelvinConv;
     }
```
Sets the ground surface temperature for LWR calculation if ground surface object is specified for an exterior surface.

Nigusse commented on src/EnergyPlus/DataSurfaces.hh:
```diff
@@ -1284,6 +1284,27 @@ namespace DataSurfaces {
+        Real64 SurfsViewFactorSum = 0.0;      // sum of view factors of ground surfaces seen by an exterior surface
+        bool IsGroundViewFactorSet = false;   // true if the ground view factor field is not blank
+    };
+
```
New structure for ground surfaces property object for use with LWR and ground surfaces solar reflectance calculation.

Nigusse commented on src/EnergyPlus/DataSurfaces.hh:
```diff
@@ -1316,14 +1338,12 @@ namespace DataSurfaces {
 
         // Default Constructor
-        SurroundingSurfacesProperty() : SkyViewFactor(-1.0), SkyTempSchNum(0), GroundViewFactor(-1.0), GroundTempSchNum(0), TotSurroundingSurface(0)
+        SurroundingSurfacesProperty() : SkyViewFactor(-1.0), SkyTempSchNum(0), TotSurroundingSurface(0)
```
Removed member variables used with the two removed fields.

Nigusse commented on src/EnergyPlus/DataSurfaces.hh:
```diff
@@ -1509,6 +1530,11 @@ struct SurfacesData : BaseGlobalStruct
+    Array1D<int> GroundSurfsPropertyNum;            // index to a ground surfaces list (defined in SurfaceProperties::GroundSurfaces)
+    Array1D<bool> UseSurfPropertyGndSurfTemp;       // true if at least one ground surface temperature schedules is specified
+    Array1D<bool> UseSurfPropertyGndSurfRefl;       // true if at least one ground surfaces reflectance schedule is specified
+    Array1D<Real64> GndReflSolarRad;                // ground surface reflected solar radiation on exterior surfaces
```
Surface variables for use exterior surface and multiple ground surfaces property model.

Myoldmopar replied:
So all 5 of these new `Array1D` have to be allocated and managed for each surface in the model?  Why not just add scalar variables to the surface data struct?  Any reason this is better?

Nigusse commented on src/EnergyPlus/HeatBalFiniteDiffManager.cc:
```diff
@@ -1573,14 +1573,21 @@ namespace HeatBalFiniteDiffManager {
+            } else {
+                Tgndsurface = state.dataMstBal->TempOutsideAirFD(Surf);
+            }
+
```
If multiple ground surfaces property object is specified, then sets ground surfaces object ground surface temperature as boundary condition instead of the default.

Nigusse commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -307,42 +307,71 @@ void InitSurfaceHeatBalance(EnergyPlusData &state)
-                    SrdSurfsViewFactor += state.dataSurface->SurroundingSurfsProperty(SrdSurfsNum).SkyViewFactor;
-                }
-                if (state.dataSurface->SurroundingSurfsProperty(SrdSurfsNum).GroundViewFactor >= 0) {
-                    SrdSurfsViewFactor += state.dataSurface->SurroundingSurfsProperty(SrdSurfsNum).GroundViewFactor;
```
Exterior surface ground view factor is now obtained from a new object. See below the changes.

Nigusse commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -307,42 +307,71 @@ void InitSurfaceHeatBalance(EnergyPlusData &state)
+                    }
+                }
+                if (SetGroundViewFactorObject) {
+                    ReSetGroundSurfacesViewFactor(state, SurfNum);
```
If the ground view factor field is blank and the ground view factor recalculated based on view factor identity, then the calculated view factor needed to be passed back to the new object. The function `ReSetGroundSurfacesViewFactor()` updates the new object variables.

Nigusse commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -2518,6 +2549,7 @@ void InitSolarHeatGains(EnergyPlusData &state)
 
             state.dataSurface->SurfSkySolarInc(SurfNum) = 0.0;
             state.dataSurface->SurfGndSolarInc(SurfNum) = 0.0;
+            state.dataSurface->GndReflSolarRad(SurfNum) = 0.0;
```
A new surface variable `GndReflSolarRad` holds the ground reflected solar radiation for each exterior surface. Now the ground surface reflectance can be different for each exterior surface; hence, the global ground reflected solar radiation variable (`state.dataEnvrn->GndSolarRad`) cannot be used.

Nigusse commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -2651,15 +2683,23 @@ void InitSolarHeatGains(EnergyPlusData &state)
+            } else {
+                GndReflSolarRad = state.dataEnvrn->GndSolarRad;
+            }
+            state.dataSurface->SurfGndSolarInc(SurfNum) = GndReflSolarRad * Surface(SurfNum).ViewFactorGround;
```
The ground reflected solar radiation for each exterior surface is calculated here and set to `GndReflSolarRad` new surface variable.

Nigusse commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -2669,10 +2709,16 @@ void InitSolarHeatGains(EnergyPlusData &state)
+                    GndSurfReflectance = state.dataSurface->GroundSurfsProperty(state.dataSurface->GroundSurfsPropertyNum(SurfNum)).SurfsReflAvg;
+                } else {
+                    GndSurfReflectance = state.dataEnvrn->GndReflectance;
+                }
```
Ground reflectance value from the new Ground surfaces object `SurfaceProperty:GroundSurfaces` is used if specified, instead of the default global.

Nigusse commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -2791,6 +2837,12 @@ void InitSolarHeatGains(EnergyPlusData &state)
+                    GndSurfReflectance = state.dataSurface->GroundSurfsProperty(state.dataSurface->GroundSurfsPropertyNum(SurfNum)).SurfsReflAvg;
+                } else {
+                    GndSurfReflectance = state.dataEnvrn->GndReflectance;
+                }
```
Ground reflectance value from the new Ground surfaces object `SurfaceProperty:GroundSurfaces` is used if specified, instead of the default global.

Nigusse commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -2822,6 +2873,14 @@ void InitSolarHeatGains(EnergyPlusData &state)
+                } else {
+                    GndSurfReflectance = state.dataEnvrn->GndReflectance;
+                }
+
```
Ground reflectance value from the new Ground surfaces object `SurfaceProperty:GroundSurfaces` is used if specified, instead of the default global.

Nigusse commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -2919,10 +2977,14 @@ void InitSolarHeatGains(EnergyPlusData &state)
+            Real64 GndReflSolarRad = state.dataEnvrn->GndSolarRad;
+            if (state.dataSurface->UseSurfPropertyGndSurfRefl(SurfNum)) {
+                GndReflSolarRad = state.dataSurface->GndReflSolarRad(SurfNum);
+            }
```
Resets the ground surfaces reflected solar radiation if `SurfaceProperty:GroundSurfaces` object is specified for a given exterior surface.

Nigusse commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -6136,6 +6198,9 @@ void CalcHeatBalanceOutsideSurf(EnergyPlusData &state,
 
+    // set ground surfaces average temperature
+    GetGroundSurfacesTemperatureAverage(state);
+
```
This function calculates ground surfaces average temperature. The ground surfaces average temperature is view factors weighted ground surfaces temperature for use with exterior surface LWR exchange.

Nigusse commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -8500,9 +8565,9 @@ void CalcOutsideSurfTemp(EnergyPlusData &state,
-        }
+    }
+    if (state.dataSurface->UseSurfPropertyGndSurfTemp(SurfNum)) {
+        TGround = state.dataSurface->GroundSurfsProperty(state.dataSurface->GroundSurfsPropertyNum(SurfNum)).SurfsTempAvg;
```
Resets the ground surface temperature if `SurfaceProperty:GroundSurfaces` is specified.

Nigusse commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -8874,4 +8939,100 @@ void GatherComponentLoadsSurfAbsFact(EnergyPlusData &state)
+        }
+        GndSurfsProperty.SurfsTempAvg = root_4(GndSurfaceTempSum / GndSurfsProperty.SurfsViewFactorSum) - DataGlobalConstants::KelvinConv;
+    }
+}
```
Function that calculates view factors weighted ground surfaces average temperature.

Myoldmopar replied:
I saw above that we are adding the Kelvin conversion back on when doing radiation calculations, as expected.  Is it reasonable to just keep the `SurfsTempAvg` variable in Kelvin and not have to subtract and add the conversion back on?  Is it used for a user facing aspect where Celsius is preferred?

Nigusse commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -8874,4 +8939,100 @@ void GatherComponentLoadsSurfAbsFact(EnergyPlusData &state)
+        }
+        GndSurfsProperty.SurfsReflAvg = GndSurfsReflSum / GndSurfsProperty.SurfsViewFactorSum;
+    }
+}
```
Function that calculates view factors weighted ground surfaces average reflectance.

Nigusse commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -8874,4 +8939,100 @@ void GatherComponentLoadsSurfAbsFact(EnergyPlusData &state)
+        return;
+    }
+    GndSurfsProperty.GndSurfs(1).ViewFactor = GndSurfsProperty.SurfsViewFactorSum;
+}
```
Function that resets the ground view factors if the ground view factor fields are blank.

Nigusse commented on src/EnergyPlus/HeatBalanceSurfaceManager.hh:
```diff
@@ -199,6 +199,12 @@ namespace HeatBalanceSurfaceManager {
+    void GetGroundSurfacesReflectanceAverage(EnergyPlusData &state);
+
+    void ReSetGroundSurfacesViewFactor(EnergyPlusData &state, int const SurfNum);
+
```
Three new functions added for use with ground surfaces property object.

Nigusse commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -8404,15 +8404,19 @@ void CalcInteriorSolarDistribution(EnergyPlusData &state)
+                        Real64 GndReflSolarRad = state.dataEnvrn->GndSolarRad;
+                        if (state.dataSurface->UseSurfPropertyGndSurfRefl(SurfNum2)) {
+                            GndReflSolarRad = state.dataSurface->GndReflSolarRad(SurfNum2);
+                        }
```
Sets the ground surfaces reflected solar radiation for each exterior surface if specified, instead of using the global ground reflected solar radiation.

Nigusse commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -8445,8 +8449,12 @@ void CalcInteriorSolarDistribution(EnergyPlusData &state)
                         Real64 DifSolarInc = state.dataEnvrn->DifSolarRad * state.dataSolarShading->SurfAnisoSkyMult(SurfNum) +
-                                             state.dataEnvrn->GndSolarRad * state.dataSurface->Surface(SurfNum).ViewFactorGround +
+                                             GndReflSolarRad * state.dataSurface->Surface(SurfNum).ViewFactorGround +
                                              ShelfSolarRad * state.dataDaylightingDevicesData->Shelf(ShelfNum).ViewFactor;
```
Same change as above.

Nigusse commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -9615,8 +9623,11 @@ void WindowShadingManager(EnergyPlusData &state)
+                if (state.dataSurface->UseSurfPropertyGndSurfRefl(ISurf)) {
+                    GndReflSolarRad = state.dataSurface->GndReflSolarRad(ISurf);
+                }
+                SolarOnWindow = BeamSolarOnWindow + SkySolarOnWindow + GndReflSolarRad * state.dataSurface->Surface(ISurf).ViewFactorGround;
```
Same change as above.

Nigusse commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -10444,8 +10455,11 @@ void SkyDifSolarShading(EnergyPlusData &state)
+        if (state.dataSurface->IsSurfPropertyGndSurfacesDefined(SurfNum)) {
+            int GndSurfsNum = state.dataSurface->GroundSurfsPropertyNum(SurfNum);
+            if (state.dataSurface->GroundSurfsProperty(GndSurfsNum).IsGroundViewFactorSet) {
+                state.dataSurface->Surface(SurfNum).ViewFactorGroundIR *= state.dataSurface->GroundSurfsProperty(GndSurfsNum).SurfsViewFactorSum;
```
The ground view factor obtained from the new object is set here if the ground view factor input field is blank.

Nigusse commented on src/EnergyPlus/SurfaceGeometry.cc:
```diff
@@ -974,6 +974,11 @@ namespace SurfaceGeometry {
+        state.dataSurface->GroundSurfsPropertyNum.allocate(state.dataSurface->TotSurfaces);
+        state.dataSurface->UseSurfPropertyGndSurfTemp.allocate(state.dataSurface->TotSurfaces);
+        state.dataSurface->UseSurfPropertyGndSurfRefl.allocate(state.dataSurface->TotSurfaces);
+        state.dataSurface->GndReflSolarRad.allocate(state.dataSurface->TotSurfaces);
```
sizes the new variables.

Nigusse commented on src/EnergyPlus/SurfaceGeometry.cc:
```diff
@@ -990,6 +995,11 @@ namespace SurfaceGeometry {
+            state.dataSurface->GroundSurfsPropertyNum(SurfNum) = 0;
+            state.dataSurface->UseSurfPropertyGndSurfTemp(SurfNum) = false;
+            state.dataSurface->UseSurfPropertyGndSurfRefl(SurfNum) = false;
+            state.dataSurface->GndReflSolarRad(SurfNum) = 0.0;
```
Initializes the new vars.

Nigusse commented on src/EnergyPlus/SurfaceGeometry.cc:
```diff
@@ -2829,6 +2839,8 @@ namespace SurfaceGeometry {
         SetupEnclosuresAndAirBoundaries(state, state.dataViewFactor->EnclRadInfo, SurfaceGeometry::enclosureType::RadiantEnclosures, ErrorsFound);
 
+        GetSurfaceGroundSurfsData(state, ErrorsFound);
+
```
Get input function for the new object.

Nigusse commented on src/EnergyPlus/SurfaceGeometry.cc:
```diff
@@ -8090,6 +8102,22 @@ namespace SurfaceGeometry {
+                    } else {
+                        state.dataSurface->SurfLocalEnvironment(Loop).GroundSurfsPtr = GndSurfsNum;
+                    }
+                }
```
Reads the new input field added to `SurfaceProperty:LocalEnvironment' object.

Nigusse commented on src/EnergyPlus/SurfaceGeometry.cc:
```diff
@@ -8108,6 +8136,12 @@ namespace SurfaceGeometry {
+                        state.dataSurface->UseSurfPropertyGndSurfTemp(SurfLoop) = true;
+                        state.dataSurface->UseSurfPropertyGndSurfRefl(SurfLoop) = true;
+                        state.dataSurface->GroundSurfsPropertyNum(SurfLoop) = state.dataSurface->SurfLocalEnvironment(Loop).GroundSurfsPtr;
+                    }
```
if the input field `Ground Surfaces Object Name` is not blank, then sets the new surface variables to `true` and sets the index to the new object.

Nigusse commented on src/EnergyPlus/SurfaceGeometry.cc:
```diff
@@ -8186,38 +8220,28 @@ namespace SurfaceGeometry {
-                if (!state.dataIPShortCut->lAlphaFieldBlanks(3)) {
-                    state.dataSurface->SurroundingSurfsProperty(Loop).GroundTempSchNum = GetScheduleIndex(state, state.dataIPShortCut->cAlphaArgs(3));
-                }
-
```
No more needed, removed fields.

Nigusse commented on src/EnergyPlus/SurfaceGeometry.cc:
```diff
@@ -8186,38 +8220,28 @@ namespace SurfaceGeometry {
-                    SurfTempArg = SurfLoop * 2 + 3; // A5, A7, A9, ...
+                    SurfNameArg = SurfLoop * 2 + 1; // A3, A5, A7, ...
+                    SurfVFArg = SurfLoop + 1;       // N2, N3, N4, ...
+                    SurfTempArg = SurfLoop * 2 + 2; // A4, A6, A8, ...
```
Needed to adjust the numbering due to the two fields removed (Idd change).

Nigusse commented on src/EnergyPlus/SurfaceGeometry.cc:
```diff
@@ -8228,6 +8252,88 @@ namespace SurfaceGeometry {
+            }
+        }
+    }
+
```
Get input function for `SurfaceProperty:GroundSurfaces' new object.

Nigusse commented on src/EnergyPlus/WindowComplexManager.cc:
```diff
@@ -2780,10 +2780,6 @@ namespace WindowComplexManager {
-                        if (state.dataSurface->SurroundingSurfsProperty(SrdSurfsNum).SkyViewFactor != -1) {
-                            state.dataSurface->Surface(SurfNum).ViewFactorGroundIR =
-                                state.dataSurface->SurroundingSurfsProperty(SrdSurfsNum).GroundViewFactor;
-                        }
```
This is redundant code. The ground view factor is now obtained from a new object as show below.

Nigusse commented on src/EnergyPlus/WindowComplexManager.cc:
```diff
@@ -2794,6 +2790,11 @@ namespace WindowComplexManager {
+                        int GndSurfsNum = state.dataSurface->GroundSurfsPropertyNum(SurfNum);
+                        state.dataSurface->Surface(SurfNum).ViewFactorGroundIR =
+                            state.dataSurface->GroundSurfsProperty(GndSurfsNum).SurfsViewFactorSum;
+                    }
```
The ground view factor is set using the new object.

Nigusse commented on src/EnergyPlus/WindowEquivalentLayer.cc:
```diff
@@ -752,10 +752,6 @@ void EQLWindowSurfaceHeatBalance(EnergyPlusData &state,
-                    if (state.dataSurface->SurroundingSurfsProperty(SrdSurfsNum).SkyViewFactor != -1) {
-                        state.dataSurface->Surface(SurfNum).ViewFactorGroundIR =
-                            state.dataSurface->SurroundingSurfsProperty(SrdSurfsNum).GroundViewFactor;
-                    }
```
This is redundant code. The ground view factor is now obtained from a new object as show below.

Nigusse commented on src/EnergyPlus/WindowManager.cc:
```diff
@@ -2635,9 +2635,6 @@ namespace WindowManager {
                         }
-                        if (state.dataSurface->SurroundingSurfsProperty(SrdSurfsNum).SkyViewFactor != -1) {
-                            surface.ViewFactorGroundIR = state.dataSurface->SurroundingSurfsProperty(SrdSurfsNum).GroundViewFactor;
-                        }
```
This is redundant code. The ground view factor is now obtained from a new object as show below.

Nigusse commented on src/EnergyPlus/WindowManager.cc:
```diff
@@ -2648,6 +2645,10 @@ namespace WindowManager {
+                    if (state.dataSurface->IsSurfPropertyGndSurfacesDefined(SurfNum)) {
+                        int GndSurfsNum = state.dataSurface->GroundSurfsPropertyNum(SurfNum);
+                        surface.ViewFactorGroundIR = state.dataSurface->GroundSurfsProperty(GndSurfsNum).SurfsViewFactorSum;
+                    }
```
The ground view factor is set using the new object.

Nigusse commented on src/Transition/InputRulesFiles/Rules22-1-0-to-22-2-0.md:
```diff
@@ -0,0 +1,53 @@
+         Note: Even when the two removed fields are blank, it is still necessary to create the new object with names (F1 and F2 fields) only 
+		       and the `Ground Surface 1 View Factor` and `Ground Surface 1 Temperature Schedule Name` fields could be either left blank or ignored.
+
+# Object ...
```
Transition rules.

Nigusse commented on testfiles/CMakeLists.txt:
```diff
@@ -568,6 +568,7 @@ add_simulation_test(IDF_FILE Supermarket_SharedAirCondenser.idf EPW_FILE USA_FL_
 add_simulation_test(IDF_FILE SupplyPlenumVAV.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 add_simulation_test(IDF_FILE SurfaceGroundHeatExchanger.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 add_simulation_test(IDF_FILE SurfaceHeatSourceTerm_RefBldgSmallOfficeNew2004.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
+add_simulation_test(IDF_FILE SurfacePropGroundSurfaces_LWR.idf EPW_FILE USA_AZ_Phoenix-Sky.Harbor.Intl.AP.722780_TMY3.epw)
```
New example file demonstrating the use of multiple ground surfaces.

Nigusse commented on testfiles/SurfacePropTest_SurfLWR.idf:
```diff
@@ -2262,8 +2262,6 @@
     0.0,                     !- Sky View Factor
     ,                        !- Sky Temperature Schedule Name
-    0.0,                     !- Ground View Factor
-    ,                        !- Ground Temperature Schedule Name
```
Deleted removed fields from an existing example idf. Manual Transition.

Nigusse commented on testfiles/SurfacePropTest_SurfLWR.idf:
```diff
@@ -2272,8 +2270,6 @@
     0.1,                     !- Sky View Factor
     ,                        !- Sky Temperature Schedule Name
-    0.0,                     !- Ground View Factor
-    ,                        !- Ground Temperature Schedule Name
```
Deleted removed fields from an existing example idf.

Nigusse commented on testfiles/SurfacePropTest_SurfLWR.idf:
```diff
@@ -2285,12 +2281,15 @@
     0.5,                     !- Sky View Factor
     ,                        !- Sky Temperature Schedule Name
-    0.3,                     !- Ground View Factor
-    ,                        !- Ground Temperature Schedule Name
```
Deleted removed fields from an existing example idf.

Nigusse commented on testfiles/SurfacePropTest_SurfLWR.idf:
```diff
@@ -2285,12 +2281,15 @@
+    GndSurfs:Zn005:Wall002:Win001,  !- Name
+    Gnd Surfs Wall002:Win001 Name,  !- Ground Surface 1 Name
+    0.3;                     !- Ground Surface 1 View Factor
+
```
Added new object to replace removed fields for an existing example idf. This is manual transition.

Nigusse commented on testfiles/SurfacePropTest_SurfLWR.idf:
```diff
@@ -2303,7 +2302,8 @@
     SrdSurfs:Zn005:Wall002:Win001,  !- Surrounding Surfaces Object Name
-    ;                        !- Outdoor Air Node Name
+    ,                        !- Outdoor Air Node Name
+    GndSurfs:Zn005:Wall002:Win001;  !- Ground Surfaces Object Name
```
This is manual transition. Added the name of the new object to the new field added to `SurfaceProperty:LocalEnvironment` object.

Nigusse commented on tst/EnergyPlus/unit/ICSCollector.unit.cc:
```diff
@@ -140,6 +140,12 @@ TEST_F(EnergyPlusFixture, ICSSolarCollectorTest_CalcPassiveExteriorBaffleGapTest
+    state->dataSurface->UseSurfPropertyGndSurfTemp.allocate(NumOfSurf);
+    state->dataSurface->UseSurfPropertyGndSurfTemp(SurfNum) = false;
+    state->dataSurface->UseSurfPropertyGndSurfRefl.allocate(NumOfSurf);
+    state->dataSurface->UseSurfPropertyGndSurfRefl(SurfNum) = false;
```
Fixed a failed unit test due to the new feature code change.

Nigusse commented on tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:
```diff
@@ -8760,3 +8763,207 @@ TEST_F(EnergyPlusFixture, SurfaceGeometry_ZoneAndSpaceAreas)
+    EXPECT_EQ("GNDSURFS LAKEAREA", GndSurfsProperty.GndSurfs(3).Name);
+    EXPECT_DOUBLE_EQ(0.1, GndSurfsProperty.GndSurfs(3).ViewFactor);
+    EXPECT_DOUBLE_EQ(0.4, GndSurfsProperty.SurfsViewFactorSum);
+}
```
New unit test added.

Nigusse commented on tst/EnergyPlus/unit/WindowManager.unit.cc:
```diff
@@ -2681,6 +2680,11 @@ TEST_F(EnergyPlusFixture, WindowManager_SrdLWRTest)
+                          "  GndSurfs:Living:North,       !-Name",
+                          "  GndSurfs North,              !-Ground Surface 1 Name",
+                          "  0.1,                         !-Ground Surface 1 View Factor",
+                          "  Ground Temp Sch;             !-Ground Surface 1 Temperature Schedule Name",
```
Transitioned unit test.

Nigusse committed:
Cleaned new object Idd

Nigusse committed:
Merged in develop; resolved conflict

Nigusse commented:
The initial commit diffs observed in `Fenestration_RefBldgSmallHotelNew2004_Chicago.idf` example is gone now. But diffs in example file `SurfacePropTest_SurfLWR.idf` is expected due to the two new report variables added.

Nigusse commented:
The four example files failed in window machine are not related to this change. They appear to be path related issue that has been addressed and merged today. I will fix the merge conflict along with one minor edit in the idd of the new object.

Nigusse committed:
Added new output variables to OutputChanges file

Nigusse committed:
Do Not Set Output Vars when input fields are blank

Nigusse committed:
Added new report vars to IO Ref Docs

Nigusse committed:
Updated Version No of new example file and resolved merge conflict

Nigusse commented:
Modified the code such that ground surfaces average temperature and average reflectance report variables will not be generated if the ground surfaces temperature and reflectance schedule name fields are blank. When these fields are blank the EnergyPlus default ground temperature and ground reflectance values are used.

Nigusse commented:
The new report variables added are:

```
Zone,Average,Surfaces Property Ground Surfaces Average Temperature [C]
Zone,Average,Surfaces Property Ground Surfaces Average Reflectance []
```

Nigusse committed:
Modified logical statement to remove warning message

Nigusse commented:
Diffs is expected in one example file `SurfacePropTest_SurfLWR.idf` only. The diffs is caused due to addition of one new idd object for transitioning the file. The diffs are in the eplustbl.htm output files. See below the diffs.
 
![image](https://user-images.githubusercontent.com/5805126/172845146-003f7f9b-958d-4dd3-b6ba-4739aa64f389.png)

This branch has a new example idf and unit tests. It is ready for review.

Nigusse committed:
Merged In Develop; Resolved Conflicts

Nigusse committed:
Vars scope cleanup

Nigusse committed:
format error statement

Nigusse committed:
Index scope cleanup

Nigusse committed:
more format error statement

Nigusse committed:
Code cleanup for readability

Nigusse committed:
Merge branch 'develop' into MultipleGroundTemperatureReflectance

Nigusse committed:
Applied Clang Format

Nigusse committed:
More cleanup

Nigusse committed:
Surrounding Surfs getInput Cleanup

Nigusse committed:
Deleted default constructs

Nigusse committed:
Merge branch 'develop' into MultipleGroundTemperatureReflectance

Nigusse commented:
Only one example file `SurfacePropTest_SurfLWR.idf` has expected diffs. The diffs are in `eplustbl` output files caused by changes in the idd.

This is ready for review.

Myoldmopar commented:
But this still needs the actual transition code, right?  Maybe I'm just missing something.

Nigusse commented:
Yes, it does need transition code. There is one example file in the test suite that need transition and I updated it manually. I also added the transition rules.

Nigusse committed:
Merged in develop; resolved conflict

Myoldmopar committed:
Merge develop, add transition rule code

Myoldmopar committed:
Merge branch 'MultipleGroundTemperatureReflectance' of github.com:NREL/EnergyPlus into MultipleGroundTemperatureReflectance

Myoldmopar committed:
Fix merge issue

Myoldmopar committed:
Fix suffix on ground surface name

Myoldmopar commented:
Well sheesh that transition rule was harder than it should have been.  But I think I have it working.  @Nigusse I'll need you to look it over and make sure it is converting things reasonably.  @mjwitte I also wouldn't mind you taking a quick peek either and give me some feedback on the transition changes.  The resulting diff looks like this:

![image](https://user-images.githubusercontent.com/849824/175079608-7fbd336f-3498-48a7-8ae4-c7fb3dd9343a.png)

Basically, for each SurroundingSurfaces object:
 - The SurroundingSurfaces object has the Ground View Factor and Ground Temperature Schedule fields removed and stored.
 - The LocalEnvironment objects are searched for a match, and when the match is found, a new Ground Surfaces Object Name is generated and added to the end of the LocalEnvironment object.
 - Finally, a new GroundSurfaces object is written using that generated name, a ground surface name generated with a "-1" suffix, and the two fields we saved from the SurroundingSurfaces object.

Does that sound/look right?  I pulled the 22.1 SurfacePropTest IDF and transitioned it up and ran it with 22.2, and it runs fine without complaining.

Nigusse commented on src/Transition/CreateNewIDFUsingRulesV22_2_0.f90:
```diff
@@ -422,8 +424,57 @@ SUBROUTINE CreateNewIDFUsingRules(EndOfFile,DiffOnly,InLfn,AskForInput,InputFile
+
+                    EXIT
+                  ENDIF
+                ENDDO
```
@Myoldmopar The transitioned objects look correct. I suggest one change, do nothing if the 4th field (`Surrounding Surfaces Object Name`) is blank. You may need to add an if block.

```           
           IF (InArgs(4) /= Blank) THEN
                matchedSurroundingName = MakeUPPERCase(InArgs(4))
                ...
                DO Num1=1,NumIDFRecords
                ENDDO

            ENDIF
```

Myoldmopar replied:
Added the check, built and tested with the 22.1 file again and everything still seems fine.  Thanks @Nigusse

Myoldmopar committed:
Only updated LocalEnvironment if SurroundingSurfaces field is not blank

Nigusse committed:
Reverted IDD and get inputs

Nigusse committed:
Updated Code and Example Files

Nigusse committed:
Fixed failed unit test due to IDD change

Nigusse committed:
Restored current method and unit test cleanup

Nigusse commented:
@Myoldmopar @mjwitte Since the idd change is rolled back, transition is not needed. But the documentation changes needs to be modified to reflect the Idd roll back and explain when the new object is used.

Nigusse committed:
Changed tab to space in an idf

Nigusse committed:
Applied clang formats

Nigusse committed:
Updated input-output documentation

Nigusse committed:
Removed transition rules

Nigusse commented:
@Myoldmopar I am finalizing the code and documentation changes. Transition is not required for this feature any more. Do you want me to remove the transition Fortran code you added for this feature?

Myoldmopar commented:
@Nigusse yeah go ahead, I'm offline for a bit this morning but I'll be back on soon. Thanks

Nigusse committed:
Removed transition code (PR#9409)

Nigusse commented:
OK. I will push the updated branch shortly.

mjwitte commented:
@Nigusse Is this ready for final review?

Nigusse commented:
@mjwitte  Yes, it is ready. Thanks.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into MultipleGroundTemperatureReflectance

Myoldmopar commented on doc/engineering-reference/src/solar-radiation-reflected-from-exterior/diffuse-reflection-of-beam-solar-and-sky.tex:
```diff
@@ -130,6 +130,32 @@ \subsection{Sky Solar Radiation Diffusely Reflected from the Ground}\label{sky-s
+\begin{array}{l}
+{QRadSWOutIncSkyDiffReflGnd(RecSurfNum)} = \\{DifSolarRad * {\rho_{gnd,avg}} * SurfReflFacSkySolGnd(RecSurfNum)~ (W/m^{2})}
+\end{array}
+\end{equation}
```
I know this was essentially copied from above, but whenever possible, try to simplify the equations down to symbols and use the `where:` approach to define the equations. Using the long variable names in the equations makes them so much harder to grok.

Myoldmopar commented on idd/Energy+.idd.in:
```diff
@@ -18787,6 +18791,113 @@ SurfaceProperty:SurroundingSurfaces,
+       \required-field
+       \type alpha
+   N1, \field Ground Surface 1 View Factor
+	   \type real
```
There seem to be a bunch of tab characters in the IDD now...I thought I had a custom-check test to catch those.

Myoldmopar replied:
So I looked, and my script only checks for stray tabs in the IDF files, not the IDD.  I modified the script to search the IDD as well, got error messages properly, and fixed the bad tabs here to get it clean again.

Nigusse replied:
I have cleaned a few tabs in the idd and documentations when I got warning error message from the regression runs.

Myoldmopar commented on scripts/dev/generate_epJSON_schema/modify_schema.py:
```diff
@@ -170,6 +170,7 @@ def isInt(s):
     'Foundation:Kiva': 'blocks',
     'SurfaceProperty:ExposedFoundationPerimeter': 'surfaces',
     'SurfaceProperty:SurroundingSurfaces': 'surfaces',
+	'SurfaceProperty:GroundSurfaces': 'ground_surfaces',
```
Another tab...I'd check your editor for why it is putting in tabs instead of spaces.  I'll fix it here since I already have the IDD tab fixes done locally.

Myoldmopar commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -8847,49 +8885,189 @@ void InitSurfacePropertyViewFactors(EnergyPlusData &state)
+    for (int SurfNum = 1; SurfNum <= state.dataSurface->TotSurfaces; ++SurfNum) {
+        if (!state.dataSurface->IsSurfPropertyGndSurfacesDefined(SurfNum)) continue;
+        auto &GndSurfsProperty = state.dataSurface->GroundSurfsProperty(state.dataSurface->GroundSurfsPropertyNum(SurfNum));
+        if (GndSurfsProperty.SurfsViewFactorSum == 0.0) {
```
Does this have to be checked every time?  Feels like something that could be determined once to make this function lighter.

Nigusse replied:
Yes, the ground surface temperature is updated every time step because it is uses schedule input.

Myoldmopar commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -8847,49 +8885,189 @@ void InitSurfacePropertyViewFactors(EnergyPlusData &state)
+        GndSurfaceTempSum = 0.0;
+        for (int gSurfNum = 1; gSurfNum <= GndSurfsProperty.NumGndSurfs; gSurfNum++) {
+            GndSurfViewFactor = GndSurfsProperty.GndSurfs(gSurfNum).ViewFactor;
+            if (GndSurfViewFactor == 0.0) continue;
```
I was going to suggest that this code is better if you just let the 0.0 pass through and the calculation work itself out, but if it is 0.0, it does avoid the Schedule lookup, so it's worth it to jump here.

Myoldmopar commented on src/EnergyPlus/DataSurfaces.hh:
```diff
@@ -1284,52 +1284,58 @@ namespace DataSurfaces {
+        // Members
+        std::string Name;                     // name of multiple ground surfaces object
+        int NumGndSurfs = 0;                  // number of groundSurfaces
+        Array1D<GroundSurfacesData> GndSurfs; // ground surfaces data
```
If you are creating structures from scratch it would be better to just use raw `std::vector` or at least `EPVector`.  We're always on a mission to reduce our dependence on `Array1D` for simple list use cases.

Myoldmopar commented on src/EnergyPlus/DataSurfaces.hh:
```diff
@@ -1780,6 +1792,7 @@ struct SurfacesData : BaseGlobalStruct
     EPVector<DataSurfaces::SurfaceLocalEnvironment> SurfLocalEnvironment;
     EPVector<DataSurfaces::SurroundingSurfacesProperty> SurroundingSurfsProperty;
     EPVector<DataSurfaces::IntMassObject> IntMassObjects;
+    EPVector<DataSurfaces::GroundSurfacesProperty> GroundSurfsProperty;
```
EPVector is fine here, although whenever possible, I would suggest pushing toward the raw vector for new code.

Nigusse replied:
@Myoldmopar I will address your comments but I don't have additional changes on my side.

Myoldmopar reviewed (commented):
I've made a few comments in here.  In general I wish some of the new code had used preferred patterns like standard library containers, or at least EPVector, where possible.  I'm not sure it's worth holding up the PR for it, since it is adding a drop to the ocean.  I also found a bunch of tabs, so I removed them and fixed up my custom-check for tabs.  I guess it's probably ready to go in but if more code changes are necessary, I'd say consider moving some of the new Array1D to at least EPVector.

I'll push my changes up now, does anyone else have anything to add here?

Myoldmopar committed:
Fix tabs in code

Nigusse committed:
Merge branch 'develop' into MultipleGroundTemperatureReflectance

Nigusse committed:
Moved GndReflSolarRad as Surface member Var

Nigusse committed:
Moved IsSurfPropertyGndSurfacesDefined var

Nigusse committed:
Moved UseSurfPropertyGndSurfTemp var

Nigusse committed:
Moved UseSurfPropertyGndSurfRefl var

Nigusse committed:
Renamed GroundSurfsPropertyNum

Nigusse committed:
Changed Array1D to EPVector

Nigusse committed:
Moved SurfHasSurroundingSurfProperties var

Nigusse committed:
Renamed SurfHasSurroundingSurfProperties

Nigusse committed:
Moved more variables to Surface Object

Nigusse committed:
Cleanup

Nigusse committed:
Applied clang format

Myoldmopar committed:
Merge branch 'MultipleGroundTemperatureReflectance' of github.com:NREL/EnergyPlus into MultipleGroundTemperatureReflectance

Myoldmopar commented:
I pulled develop and the latest changes into my local branch.  Everything builds and passes fine.  I suspect this is good to go, but I'll let CI run a bit before merging.  Thanks @Nigusse

Myoldmopar commented:
This looks perfect.  Integration coverage is highly likely to be fine, so I'm going to merge this.  Thanks @Nigusse !
